### ISSUE TITLE
about alignment

### ISSUE THREAD

**Issue by MorganCZY** [begeekmyfriend/tacotron2#22]
I trained this repo with BiaoBei dataset and no parameters have been modified. However, it didn't get the perfect alignment even after 278 epoches. Compared to your training [situation](https://github.com/begeekmyfriend/tacotron2/issues/4#issuecomment-564454381), would you please give me some suggestions on the improvement of alignment.
![image](https://user-images.githubusercontent.com/33675916/81381434-4e3fe980-913f-11ea-928a-19d2a68f8850.png)

**Comment by begeekmyfriend** [comment]
Well it is strange for both of us. If you enabled `amp_run` in parsing arguments, please set amp operational level into `O0` to rise up the training precision. If not, you might try the old Mel spectrograms extraction method that it would load Mel spectrograms from numpy files as this code [shows](https://github.com/begeekmyfriend/tacotron2/blob/master/tacotron2/data_function.py#L72). Let us make some comparison.

**Comment by MorganCZY** [comment]
I didn't activate any options about amp and distributed_run. I'm gonna compare two mechnism of calculating mel spectrogram. Meanwhile, if you have found any clues, pleaes share them out asap. Thanks in advance!

**Comment by MorganCZY** [comment]
I retrained this repo with the option "load_mel_from_disk", but the alignment is still not correct after 30 epochs.
![image](https://user-images.githubusercontent.com/33675916/81465702-73dff800-91fe-11ea-83d6-df84eb7cd4fe.png)

**Comment by begeekmyfriend** [comment]
I forgot to tell you that you need to change the [mel_pad_val](https://github.com/begeekmyfriend/tacotron2/blob/master/tacotron2/loader.py#L43) back into -4 or -5 which should be the lowest Mel value.

**Comment by MorganCZY** [comment]
I noticed this param, but I have a question to consult. This param is just used as the value of silence when padding. Based on my knowladge, a value that doesn't appear in the numerical range mel is usually chooesed.  Then why will -11.5129, which is far smaller than -4 or -5, cause alignment issue? (Btw, I will immediately retrain this repo with modifying this param to -4 to check.)

**Comment by MorganCZY** [comment]
After carefully check your codes, I'm supposed to simply change mel_pad_val to -4(or the lowest mel value). Because, in your repo, mel-spec is normalized by the function "[dynamic_range_compression](https://github.com/begeekmyfriend/tacotron2/blob/da23553b17e6f60ac87317458e916b2c57673e8d/common/audio_processing.py#L112)" in the mode of "load_mel_from_disk=false", after which the minimum of mel is -11.5129(equals to torch.log(1e-5)). Thus, the final mel-specs fed into tacotron2 are in numerical range of [-11.5129, torch.log(mel.max())], which is not a symmetrical interval. I'm not sure whether it it the reason to ruin the training. I wonder when training this tacotron2, is a symmetrical interval of mel value needed?

**Comment by begeekmyfriend** [comment]
I have not tested your idea closely, but the implementation in `stft.py` is used in many other TTS projects like `MelGAN`, `WaveGlow` and so on where the Mel values are not symmetrical and the lowest value is -11.5129. But I have no idea whether it ruins the alignment in TTS training.

**Comment by MorganCZY** [comment]
I retrained the repo with mel-specs calculated in the mode "load_mel_from_disk=False", which calls the [line](https://github.com/begeekmyfriend/tacotron2/blob/da23553b17e6f60ac87317458e916b2c57673e8d/common/preprocessor.py#L78). The alignment is then much more normal. However, most of them are as the the first following picture. Some of them are perfect as the second following picture. I couldn't understand the reason for the first kind of alignment. BEG for your suggestions!
![image](https://user-images.githubusercontent.com/33675916/81519131-bd515400-9372-11ea-87ff-af2184478f56.png)
![image](https://user-images.githubusercontent.com/33675916/81519142-c93d1600-9372-11ea-9b47-e963ce732485.png)

**Comment by begeekmyfriend** [comment]
It confused me too. But the numpy file loading approach would lead to the uncertain lowest value of GTA Mel spectrograms due to the [ref_min_db](https://github.com/begeekmyfriend/tacotron2/blob/master/common/audio.py#L60) bias in Mel extraction. And therefore it would also lead to the uncertainty of Mel padding of the vocoder due to different corpus of anchors.
There is still a way worth trying for the approach in `stft.py`, you can comment out [preemphasis](https://github.com/begeekmyfriend/tacotron2/blob/master/common/utils.py#L60) method in wav loading as well as [deemphasis](https://github.com/begeekmyfriend/tacotron2/blob/da23553b17e6f60ac87317458e916b2c57673e8d/common/audio_processing.py#L103) method in griffin-lim function which I added myself. I am   not sure if it works. By the way, I did not modify any code in `stft.py`.

**Comment by MorganCZY** [comment]
> But the numpy file loading approach would lead to the uncertain lowest value of GTA Mel spectrograms due to the [ref_min_db](https://github.com/begeekmyfriend/tacotron2/blob/master/common/audio.py#L60) bias in Mel extraction. And therefore it would also lead to the uncertainty of Mel padding of the vocoder due to different corpus of anchors.

The lowest value of GTA mel specs is -4.0, determined by codes [here](https://github.com/begeekmyfriend/tacotron2/blob/da23553b17e6f60ac87317458e916b2c57673e8d/common/audio.py#L136). Thus, I set "mel_pad_val=-4".

**Comment by begeekmyfriend** [comment]
You will know that it is not correct when you generate GTA Mel spectrograms.

**Comment by MorganCZY** [comment]
I've figured out why the alignment shows in such strange way. When [ploting alignment](https://github.com/begeekmyfriend/tacotron2/blob/da23553b17e6f60ac87317458e916b2c57673e8d/train.py#L374), a random index is first generated to decide plotting which one out of a batch alignments. In training stage, one batch data, including texts and mels are padded  based on their corresponding max lengths. A random alignment(e.g. with shape[128,151]) out of one batch(with the number of len(alignments)) has (128-seq_len) lines of zeros. Meanwhile its second dim, corresponding to decoding steps, also has some meaningless values because of padding operation [here](https://github.com/begeekmyfriend/tacotron2/blob/da23553b17e6f60ac87317458e916b2c57673e8d/tacotron2/data_function.py#L133).
I modified relevant codes as follows, and now the alignment shows correctly.
![image](https://user-images.githubusercontent.com/33675916/81673505-5751f300-947e-11ea-86c2-00350645cabb.png)

**Comment by begeekmyfriend** [comment]
Good job! I have an idea that you can try replacing [spectral_normalize](https://github.com/begeekmyfriend/tacotron2/blob/da23553b17e6f60ac87317458e916b2c57673e8d/common/layers.py#L104) with [amp_to_db](https://github.com/begeekmyfriend/tacotron2/blob/da23553b17e6f60ac87317458e916b2c57673e8d/common/audio.py#L125) to check out whether the alignment get improved.

**Comment by MorganCZY** [comment]
Actually, in my project, I have totally replaced data processing by functions in common/audio.py and set option "load_mel_from_disk"=True . It seems easier and more stable for training, at least for me.
